File: packages/sdk-ts/src/notifications.ts
// suite.notifications.* — outbox-style notifications.
//
// Endpoint paths and JSON shapes are the canonical contract from
// `apps/dashboard/src/lib/api.ts` (the Phase 10.1 Notifications
// section). Schemas below mirror the zod schemas there; if the
// dashboard contract moves, move these together.
//
// Public API stays camelCase; the wire stays snake_case. The dashboard
// schemas use snake_case already, so we translate on input/output here
// (the `from` -> `from_` mapping in particular avoids the reserved-word
// foot-gun some JS frameworks hit).

import { z } from "zod"
import { request, type HttpOptions } from "./_http.js"

// ---------- shared schemas (mirror lib/api.ts, camelCase) ----------

export const NotificationKindSchema = z.enum(["email", "sms", "push", "log"])
export type NotificationKind = z.infer<typeof NotificationKindSchema>

export const NotificationStatusSchema = z.enum([
  "queued",
  "sending",
  "sent",
  "failed",
  "skipped",
])
export type NotificationStatus = z.infer<typeof NotificationStatusSchema>

export const NotificationSchema = z.object({
  id: z.string(),
  tenantId: z.string().nullable(),
  kind: NotificationKindSchema,
  adapter: z.string().nullable(),
  template: z.string(),
  to: z.string(),
  from: z.string().nullable(),
  subject: z.string().nullable(),
  data: z.record(z.string(), z.unknown()),
  status: NotificationStatusSchema,
  providerMessageId: z.string().nullable(),
  attempts: z.number(),
  lastError: z.string().nullable(),
  scheduledAt: z.string(),
  sentAt: z.string().nullable(),
  createdAt: z.string(),
})
export type Notification = z.infer<typeof NotificationSchema>

export const NotificationListSchema = z.object({
  notifications: z.array(NotificationSchema),
  total: z.number(),
  hasMore: z.boolean(),
})
export type NotificationList = z.infer<typeof NotificationListSchema>

export const AdapterCountSchema = z.object({
  adapter: z.string(),
  count: z.number(),
})
export type AdapterCount = z.infer<typeof AdapterCountSchema>

export const NotificationStatsSchema = z.object({
  byStatus: z.record(z.string(), z.number()),
  byAdapter: z.array(AdapterCountSchema),
  sentToday: z.number(),
  failedToday: z.number(),
})
export type NotificationStats = z.infer<typeof NotificationStatsSchema>

// ---------- public API ----------

const ALLOWED_KINDS: readonly NotificationKind[] = ["email", "sms", "push", "log"]

export interface SendNotificationOptions extends HttpOptions {
  /** Recipient address. Email for kind=email, phone for sms, etc. */
  to: string
  /** Template slug. The runtime renders against `data`. Default: "default". */
  template?: string
  /** Channel. Default: "email". */
  kind?: NotificationKind | string
  /** Subject (email/push only). */
  subject?: string
  /** From address (overrides the adapter default). */
  from?: string
  /** Free-form payload passed into the template renderer. */
  data?: Record<string, unknown>
  /** ISO 8601 timestamp. Omit to send immediately. */
  scheduledAt?: string
}

export interface EmailNotificationOptions extends HttpOptions {
  /** Plain-text body. Folded into `data.body` so the default renderer can use it. */
  body?: string
  /** Template slug. Default: "default". */
  template?: string
  /** Free-form payload passed into the template renderer. */
  data?: Record<string, unknown>
  /** From address (overrides the adapter default). */
  from?: string
  /** ISO 8601 timestamp. Omit to send immediately. */
  scheduledAt?: string
}

export interface ListNotificationsOptions extends HttpOptions {
  tenant?: string
  status?: NotificationStatus | string
  kind?: NotificationKind | string
  limit?: number
  offset?: number
}

/** Queue a notification for delivery. */
export async function send(opts: SendNotificationOptions): Promise<Notification> {
  if (typeof opts.to !== "string" || opts.to.length === 0) {
    throw new Error("`to` must be a non-empty string")
  }
  const {
    to,
    template = "default",
    kind = "email",
    subject,
    from,
    data,
    scheduledAt,
    ...http
  } = opts
  if (typeof template !== "string" || template.length === 0) {
    throw new Error("`template` must be a non-empty string")
  }
  if (!(ALLOWED_KINDS as readonly string[]).includes(kind)) {
    throw new Error(
      `kind must be one of: ${ALLOWED_KINDS.join(", ")}; got ${JSON.stringify(kind)}`,
    )
  }
  const body: Record<string, unknown> = {
    to,
    template,
    kind,
  }
  if (subject !== undefined) body.subject = subject
  if (from !== undefined) body.from = from
  if (data !== undefined) body.data = { ...data }
  if (scheduledAt !== undefined) body.scheduled_at = scheduledAt

  const raw = await request<unknown>("POST", "/notifications", body, http)
  return NotificationSchema.parse(camelizeNotification(raw))
}

/**
 * Convenience for `send({ kind: "email", ... })`.
 *
 * `body` is folded into `data.body` so the default template renderer
 * (which substitutes `{{key}}` tokens against `data`) can produce a
 * plain-text email without a real template registry.
 */
export async function email(
  to: string,
  subject: string,
  opts: EmailNotificationOptions = {},
): Promise<Notification> {
  if (typeof subject !== "string" || subject.length === 0) {
    throw new Error("`subject` must be a non-empty string")
  }
  const { body, template, data, from, scheduledAt, ...http } = opts
  const payloadData: Record<string, unknown> = { ...(data ?? {}) }
  if (body !== undefined) payloadData.body = body
  return send({
    to,
    subject,
    kind: "email",
    template: template ?? "default",
    data: payloadData,
    from,
    scheduledAt,
    ...http,
  })
}

/** List notifications with optional filters. */
export async function list(
  opts: ListNotificationsOptions = {},
): Promise<NotificationList> {
  const { tenant, status, kind, limit = 50, offset = 0, ...http } = opts
  if (limit <= 0) throw new Error("limit must be a positive integer")
  if (offset < 0) throw new Error("offset must be non-negative")
  const qs = new URLSearchParams()
  if (tenant !== undefined) qs.set("tenant", tenant)
  if (status !== undefined) qs.set("status", status)
  if (kind !== undefined) qs.set("kind", kind)
  qs.set("limit", String(limit))
  qs.set("offset", String(offset))
  const raw = await request<unknown>(
    "GET",
    `/notifications?${qs.toString()}`,
    null,
    http,
  )
  return NotificationListSchema.parse(camelizeNotificationList(raw))
}

/** Fetch a single notification by id. */
export async function get(
  id: string,
  opts: HttpOptions = {},
): Promise<Notification> {
  if (typeof id !== "string" || id.length === 0) {
    throw new Error("notification id must be a non-empty string")
  }
  const raw = await request<unknown>(
    "GET",
    `/notifications/${encodeURIComponent(id)}`,
    null,
    opts,
  )
  return NotificationSchema.parse(camelizeNotification(raw))
}

/** Return KPI aggregates for the dashboard's notifications tab. */
export async function stats(
  opts: HttpOptions = {},
): Promise<NotificationStats> {
  const raw = await request<unknown>("GET", "/notifications/stats", null, opts)
  return NotificationStatsSchema.parse(camelizeStats(raw))
}

// ---------- helpers ----------

/**
 * Translate the snake_case wire envelope for a single notification into
 * the camelCase shape declared by `NotificationSchema`. `data` is
 * forwarded verbatim (it's a free-form payload).
 */
function camelizeNotification(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  return {
    id: src.id,
    tenantId: src.tenant_id ?? src.tenantId ?? null,
    kind: src.kind,
    adapter: src.adapter ?? null,
    template: src.template,
    to: src.to,
    from: src.from ?? null,
    subject: src.subject ?? null,
    data: src.data ?? {},
    status: src.status,
    providerMessageId:
      src.provider_message_id ?? src.providerMessageId ?? null,
    attempts: src.attempts,
    lastError: src.last_error ?? src.lastError ?? null,
    scheduledAt: src.scheduled_at ?? src.scheduledAt,
    sentAt: src.sent_at ?? src.sentAt ?? null,
    createdAt: src.created_at ?? src.createdAt,
  }
}

function camelizeNotificationList(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  const notifications = Array.isArray(src.notifications)
    ? src.notifications.map(camelizeNotification)
    : src.notifications
  return {
    notifications,
    total: src.total,
    hasMore: src.has_more ?? src.hasMore,
  }
}

function camelizeStats(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  return {
    byStatus: src.by_status ?? src.byStatus ?? {},
    byAdapter: src.by_adapter ?? src.byAdapter ?? [],
    sentToday: src.sent_today ?? src.sentToday ?? 0,
    failedToday: src.failed_today ?? src.failedToday ?? 0,
  }
}

/** Namespace object — the shape `suite.notifications` is built from. */
export const notifications = {
  send,
  email,
  list,
  get,
  stats,
} as const

File: packages/sdk-py/af_stack/notifications.py
# SPDX-License-Identifier: Apache-2.0

"""``suite.notifications.*`` — outbox-style notifications.

Maps to the REST surface defined in ``apps/dashboard/src/lib/api.ts``
under the Phase 10.1 Notifications section:

==================================  ====================================
SDK call                            Endpoint
==================================  ====================================
:func:`send`                        ``POST  /api/v1/notifications``
:func:`email`                       ``POST  /api/v1/notifications`` (kind=email)
:func:`list`                        ``GET   /api/v1/notifications``
:func:`get`                         ``GET   /api/v1/notifications/{id}``
:func:`stats`                       ``GET   /api/v1/notifications/stats``
==================================  ====================================

The runtime persists every send in ``suite_notifications`` at
``status='queued'`` and dispatches via a background worker through the
configured adapter (log-stub for dev, Resend for prod). The dashboard's
"Notifications" tab reads the same outbox so operators see what was
sent, queued, or failed without dipping into provider dashboards.

The :class:`Notification`, :class:`NotificationList`, and
:class:`NotificationStats` Pydantic models mirror the zod schemas in
``lib/api.ts`` exactly — keep them in sync when the contract changes.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from . import _http

NotificationKind = Literal["email", "sms", "push", "log"]
NotificationStatus = Literal["queued", "sending", "sent", "failed", "skipped"]

_ALLOWED_KINDS: tuple[str, ...] = ("email", "sms", "push", "log")


class Notification(BaseModel):
    """One notification row from the outbox.

    Mirrors ``NotificationSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    id: str
    tenant_id: str | None = None
    kind: NotificationKind
    adapter: str | None = None
    template: str
    to: str
    from_: str | None = Field(default=None, alias="from")
    subject: str | None = None
    data: dict[str, Any] = Field(default_factory=dict)
    status: NotificationStatus
    provider_message_id: str | None = None
    attempts: int = 0
    last_error: str | None = None
    scheduled_at: str
    sent_at: str | None = None
    created_at: str


class NotificationList(BaseModel):
    """Paginated list of notifications.

    Mirrors ``NotificationListSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    notifications: list[Notification] = Field(default_factory=list)
    total: int = 0
    has_more: bool = False


class AdapterCount(BaseModel):
    """One entry in :class:`NotificationStats.by_adapter`."""

    model_config = ConfigDict(extra="allow")

    adapter: str
    count: int


class NotificationStats(BaseModel):
    """KPI aggregates for the notifications dashboard.

    Mirrors ``NotificationStatsSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    by_status: dict[str, int] = Field(default_factory=dict)
    by_adapter: list[AdapterCount] = Field(default_factory=list)
    sent_today: int = 0
    failed_today: int = 0


# ─── Operations ───────────────────────────────────────────────────────────


async def send(
    *,
    to: str,
    template: str = "default",
    kind: NotificationKind | str = "email",
    subject: str | None = None,
    from_: str | None = None,
    data: dict[str, Any] | None = None,
    scheduled_at: str | None = None,
) -> Notification:
    """Queue a notification for delivery.

    The row is persisted at ``status='queued'`` and the background worker
    dispatches it through the configured adapter. The returned
    :class:`Notification` carries the queued row — poll :func:`get` or
    watch the dashboard for the terminal transition.
    """
    if not isinstance(to, str) or not to:
        raise ValueError("`to` must be a non-empty string")
    if not isinstance(template, str) or not template:
        raise ValueError("`template` must be a non-empty string")
    if kind not in _ALLOWED_KINDS:
        allowed = ", ".join(_ALLOWED_KINDS)
        raise ValueError(f"`kind` must be one of: {allowed}; got {kind!r}")

    payload: dict[str, Any] = {
        "kind": kind,
        "template": template,
        "to": to,
    }
    if subject is not None:
        payload["subject"] = subject
    if from_ is not None:
        payload["from"] = from_
    if data is not None:
        payload["data"] = dict(data)
    if scheduled_at is not None:
        payload["scheduled_at"] = scheduled_at

    body = await _http.request_json("POST", "/notifications", json=payload)
    return Notification.model_validate(body or {})


async def email(
    to: str,
    subject: str,
    *,
    body: str | None = None,
    template: str = "default",
    data: dict[str, Any] | None = None,
    from_: str | None = None,
    scheduled_at: str | None = None,
) -> Notification:
    """Convenience for ``send(kind="email")``.

    ``body`` is folded into ``data["body"]`` so the default template
    renderer (which substitutes ``{{key}}`` tokens against ``data``)
    can produce a plain-text email without a real template registry.
    """
    if not isinstance(subject, str) or not subject:
        raise ValueError("`subject` must be a non-empty string")
    payload_data: dict[str, Any] = dict(data) if data else {}
    if body is not None:
        payload_data["body"] = body
    return await send(
        kind="email",
        to=to,
        subject=subject,
        template=template,
        data=payload_data,
        from_=from_,
        scheduled_at=scheduled_at,
    )


async def list(  # noqa: A001 — mirrors the JS `notifications.list()` ergonomics
    *,
    tenant: str | None = None,
    status: NotificationStatus | str | None = None,
    kind: NotificationKind | str | None = None,
    limit: int = 50,
    offset: int = 0,
) -> NotificationList:
    """List notifications with optional filters."""
    if limit <= 0:
        raise ValueError("limit must be a positive integer")
    if offset < 0:
        raise ValueError("offset must be non-negative")
    params: dict[str, Any] = {"limit": limit, "offset": offset}
    if tenant is not None:
        params["tenant"] = tenant
    if status is not None:
        params["status"] = status
    if kind is not None:
        params["kind"] = kind
    body = await _http.request_json("GET", "/notifications", params=params)
    return NotificationList.model_validate(
        body or {"notifications": [], "total": 0, "has_more": False}
    )


async def get(id: str) -> Notification:
    """Fetch a notification by id."""
    if not id:
        raise ValueError("notification id must be a non-empty string")
    body = await _http.request_json("GET", f"/notifications/{id}")
    return Notification.model_validate(body or {})


async def stats() -> NotificationStats:
    """Return KPI aggregates for the dashboard's notifications tab."""
    body = await _http.request_json("GET", "/notifications/stats")
    return NotificationStats.model_validate(
        body or {"by_status": {}, "by_adapter": [], "sent_today": 0, "failed_today": 0}
    )


__all__ = [
    "AdapterCount",
    "Notification",
    "NotificationKind",
    "NotificationList",
    "NotificationStats",
    "NotificationStatus",
    "email",
    "get",
    "list",
    "send",
    "stats",
]